File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
private:
    void possibleCombinations(int open, int close, int n, vector<string> &combos, string &st){
        if (open == n && close == n){
            combos.push_back(st);
            return;
        }

        if (open < n){
            st += "(";
            possibleCombinations(open+1, close, n, combos, st);
            st.pop_back();
        }
        if (close < open){
            st += ")";
            possibleCombinations(open, close+1, n, combos, st);
            st.pop_back();
        }
    }
public:
    vector<string> generateParenthesis(int n) {
        vector<string> combos;
        string st;
        possibleCombinations(0, 0, n, combos, st);
        return combos;
    }
};

File: 22-generate-parentheses/generate-parentheses.go
func solve(n, open, close int, temp []byte, ans *[]string) {
    if open == n && close == n {
        s := string(temp)
        *ans = append(*ans, s)
        return
    }

    if open < n {
        temp = append(temp, '(')
        solve(n, open+1, close, temp, ans)
        temp = temp[:len(temp)-1]
    }

    if close < open {
        temp = append(temp, ')')
        solve(n, open, close+1, temp, ans)
        temp = temp[:len(temp)-1]
    }
}

func generateParenthesis(n int) []string {
    var ans []string
    var temp []byte
    solve(n, 0, 0, temp, &ans)
    return ans
}